### ISSUE TITLE
Making the experiment evaluatable

### ISSUE THREAD

**Issue by ChenStormsout** [ChenStormsout/Networked_AI_System_Course#2]
# Why  
At the end of our project, we will need certain plots or metrics, that make it possible to evaluate that our approach of Federated learning combined with Meta learning works, how well it works, and what method works best. Only by having those information, it is possible to have a good report.  
# How  
By logging the important metrics from the parameter server and the nodes, we are able to evaluate it later on. However, this is only valuable if the nodes act deterministically.  
# What  

- **Make nodes deterministic:**  
Add an environment variable to the docker compose, that sets a random seed, and add to the nodes the functionality to read this variable and set it as random seed. In case there is no random seed given, a random random seed shall be set. The random seed needs to apply to numpy and tensorflow and should be applied once at the very beginning of the start up (in app.yp)
- **Enable persisten logging:**  
The logged data should not be lost when the containers are stopped removed, we will need a persistent volume for all docker containers. It seems reasonable to share one volume for all container, which is mounted in a subfolder of our git repo.
- **Log server data:**  
The parameter server is to a certain extend the central point of information. Therefore it might be a good approach to log all information that is contained by it. Important information to be logged is for example the recieved scores, used hyper-parameters, send hyper-parameter recommendations, timestamps of the recieved update and the number of nodes communicating. The model weights are probably less interesting. There might be other information that we want to log as well. The core question is: What infomration do we need to track, to make a good evaluation in the report?
- ** Log node data: ** At the current state, there is also some information that is not sent to server. For instance, the initial random seed, the pre-training performance (Sidenote: For random neural networks weights, the pre-training performance is expected to be around 0.5 for our dataset. If federated learning has an effect, the average pre-training performance should be higher than 0.5), at what data cycle it was it was trained the xth time at what timestamp, etc. Again, the core question is:  What infomration do we need to track, to make a good evaluation in the report?

Additional information: Every node comes with a unique id generated by the uuid1() function in the node runner. It is mandatory that those ids remain unique. Please verify if that stays true, if you apply random seeding, i.e. test if two containers with the same random seed have different ids. Those ids are also valuable for logging as you can afterwards group logged information by node. Therefore, maybe use/include them for logging
